### PR TITLE
dnsmasq: add DNS-RR types (add arbitrary resource records)

### PR DESCRIPTION
**Pull request by systemcrash**
Tested on 22.03.5

Can set the review ball rolling with @pprindeville @vgaetera @jow-

### CODE REVIEW

File: package/network/services/dnsmasq/files/dnsmasq.init [hunk 1]
```diff
@@ -787,6 +787,28 @@ dhcp_hostrecord_add() {
...
+	config_get dnsrr "$cfg" dnsrr
+	[ -n "$dnsrr" ] || return 0
+
+	config_get rrnumber "$cfg" rrnumber
+	[ -n "$rrnumber" ] && [ "$rrnumber" -gt 0 ] || return 0
+
+	config_get hexdata "$cfg" hexdata
+
+	# dnsmasq accepts XX:XX, XX XX, or XXXX forms or mixtures thereof
+	if [ -n "${hexdata//[0-9a-fA-F\:\ ]/}" ]; then
```
Comment (pprindeville) [inline package/network/services/dnsmasq/files/dnsmasq.init:804]: I would use a `case` and exact globs matching those patterns.

**Comment by pprindeville** [comment]
In isc-dhcp we use explicit blocks for MX, SRV, CNAME, etc. why not do the same?

**Comment by systemcrash** [comment]
>
> I would use a case and exact globs matching those patterns.
>
"and combinations thereof" is the overarching match. Or do you imagine
something different?

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 1 above)
Comment (pprindeville) [reply package/network/services/dnsmasq/files/dnsmasq.init:804]: Specifically:

```
case $hexdata in
[0-9a-fA-F][0-9a-fA-F][ :][0-9a-fA-F][0-9a-fA-F]|[0-9a-fA-F][0-9a-fA-F][0-9a-fA-F][0-9a-fA-F])
    # handle as valid data
    ;;
*)
   # handle as an error
esac
```

**Comment by pprindeville** [comment]
> I would use a case and exact globs matching those patterns.
> "and combinations thereof" is the overarching match. Or do you imagine something different?

Like what sort of patterns?

File: package/network/services/dnsmasq/files/dnsmasq.init (hunk 1 above)
Comment (systemcrash) [reply package/network/services/dnsmasq/files/dnsmasq.init:804]: Ah, sort of how I imagined. Thank you. 

Then, I believe it is OK as is. hexdata can be much longer strings, just that each byte can be space or colon separated, or not at all. dnsmasq takes all. 

One could just accept a single format to simplify.

**Comment by systemcrash** [comment]
> In isc-dhcp we use explicit blocks for MX, SRV, CNAME, etc. why not do the same?

Could you point to specific lines? Or did you miss what this does? dnsmasq already handles those specific cases, but it also handles dns-rr for those types not yet understood by dnsmasq.

**Comment by systemcrash** [comment]
> what sort of RR’s are you anticipating?

https://forum.openwrt.org/t/resolving-query-type-65-to-local-address-for-ios-clients-in-dnsmasq/179504/11

**Comment by systemcrash** [comment]
Can you review?

@neheb 
@BKPepe

**Review by ynezz — APPROVED** [review]
> > what sort of RR’s are you anticipating?
> 
> https://forum.openwrt.org/t/resolving-query-type-65-to-local-address-for-ios-clients-in-dnsmasq/179504/11

Nice,  I would add that info into the commit description, basically never expect, that reviewer/reader of that commit has the complete context and thus provide it in the commit description, the more details, the better. Usually the answer to "Why the changes are needed?" question should be good enough.
 
So in this case, I would probably write something along this lines:

> Add support for handling of DNS RR (Resource Records) requests, which are needed for the  HTTPS Type 65 records, introduced to support the DNS-based Service Discovery (DNS-SD) mechanism for HTTPS services and defined in the RFC 9460 (9.1.  Query Names for HTTPS RRs).
>
> References: https://forum.openwrt.org/t/resolving-query-type-65-to-local-address-for-ios-clients-in-dnsmasq/179504/11

**Comment by systemcrash** [comment]
> So in this case, I would probably write something along this lines:

Suitable. Included. Also renamed `dnsrr` to `rrname` to align better with the parameters on manpage (context is more clear).

**Comment by systemcrash** [comment]
Can we get some more reviews or an approval here?

**Review by pprindeville — CHANGES_REQUESTED** [review]
Version bump on `PKG_RELEASE`?

**Comment by systemcrash** [comment]
> Version bump on `PKG_RELEASE`?

That thing is so dynamic right now, this PR will quickly get outdated :D

**Review by pprindeville — APPROVED** [review]
LGTM

**Review by GetVladimir — APPROVED** [review]

**Review by joed74 — APPROVED** [review]

**Comment by systemcrash** [comment]
Closing in favour of #14975 
Grateful if you could put your approval there @ynezz

**Pull request closed without merging**